File: metadata/mill-metadata-core/src/main/java/io/qpointz/mill/metadata/domain/FacetConverter.java
package io.qpointz.mill.metadata.domain;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import java.util.Optional;

public class FacetConverter {

    private static final FacetConverter DEFAULT = new FacetConverter(createDefaultMapper());

    private final ObjectMapper objectMapper;

    public FacetConverter(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    public static FacetConverter defaultConverter() {
        return DEFAULT;
    }

    public <T> Optional<T> convert(Object raw, Class<T> facetClass) {
        if (raw == null) {
            return Optional.empty();
        }
        if (facetClass.isInstance(raw)) {
            return Optional.of(facetClass.cast(raw));
        }
        try {
            T deserialized = objectMapper.convertValue(raw, facetClass);
            return Optional.of(deserialized);
        } catch (Exception e) {
            return Optional.empty();
        }
    }

    private static ObjectMapper createDefaultMapper() {
        ObjectMapper mapper = new ObjectMapper();
        mapper.registerModule(new JavaTimeModule());
        mapper.registerModule(new Jdk8Module());
        return mapper;
    }
}

File: metadata/mill-metadata-core/src/main/kotlin/io/qpointz/mill/metadata/domain/FacetConverter.kt
package io.qpointz.mill.metadata.domain

import io.qpointz.mill.utils.JsonUtils
import java.util.Optional

/** Converts raw facet payloads to typed facet classes using the shared Mill JSON mapper. */
class FacetConverter {

    private val objectMapper = JsonUtils.defaultJsonMapper()

    /** @param raw payload or already-converted instance */
    fun <T : Any> convert(raw: Any?, facetClass: Class<T>): Optional<T> {
        if (raw == null) return Optional.empty()
        if (facetClass.isInstance(raw)) return Optional.of(facetClass.cast(raw))
        return try {
            Optional.ofNullable(objectMapper.convertValue(raw, facetClass))
        } catch (_: Exception) {
            Optional.empty()
        }
    }

    companion object {
        private val DEFAULT = FacetConverter()

        @JvmStatic
        fun defaultConverter(): FacetConverter = DEFAULT
    }
}
